src/osd/scrubber/osd_scrub_sched.cc

```diff
@@ -0,0 +1,721 @@
+  f->dump_bool("forced", m_sched_time == PgScrubber::scrub_must_stamp());
+  f->close_section();
+}
+
+void ScrubQueue::dump_scrubs(ceph::Formatter* f) const
+{
+  ceph_assert(f != nullptr);
+  std::lock_guard l(jobs_lock);
+
+  f->open_array_section("scrubs");
+
+  for_each(to_scrub.cbegin(), to_scrub.cend(),
+	   [&f](const ScrubJobRef& j) { j->dump(f); });
+
+  for_each(penalized.cbegin(), penalized.cend(),
+	   [&f](const ScrubJobRef& j) { j->dump(f); });
+
+  f->close_section();
+}
+
+ScrubQueue::ScrubQContainer ScrubQueue::list_all_valid() const
+{
+  ScrubQueue::ScrubQContainer all_jobs;
+  all_jobs.reserve(to_scrub.size() + penalized.size());
+
+  std::unique_lock l(jobs_lock, 2'000ms);
+  ceph_assert(l.owns_lock());
+
+  std::copy_if(to_scrub.begin(), to_scrub.end(), std::back_inserter(all_jobs), valid_job);
+  std::copy_if(penalized.begin(), penalized.end(), std::back_inserter(all_jobs),
+	       valid_job);
+
+  return all_jobs;
+}
+
+// ////////////////////////////////////////////////////////////////////////////////// //
+// ScrubJob - scrub resource management
+
+bool ScrubQueue::can_inc_scrubs() const
+{
+  // consider removing the lock here. Caller already handles delayed
+  // inc_scrubs_local() failures
+  std::lock_guard l{resource_lock};
+
+  if (m_scrubs_local + m_scrubs_remote < m_cct->_conf->osd_max_scrubs) {
+    return true;
+  }
+
+  dout(20) << __func__ << " == false " << m_scrubs_local << " local + " << m_scrubs_remote
+	   << " remote >= max " << m_cct->_conf->osd_max_scrubs << dendl;
+  return false;
+}
+
+bool ScrubQueue::inc_scrubs_local()
+{
+  std::lock_guard l{resource_lock};
+
+  if (m_scrubs_local + m_scrubs_remote < m_cct->_conf->osd_max_scrubs) {
+    ++m_scrubs_local;
+    return true;
+  }
+
+  dout(20) << __func__ << " " << m_scrubs_local << " local + " << m_scrubs_remote
+	   << " remote >= max " << m_cct->_conf->osd_max_scrubs << dendl;
+  return false;
+}
+
+void ScrubQueue::dec_scrubs_local()
+{
+  dout(20) << __func__ << " " << m_scrubs_local << " -> " << (m_scrubs_local - 1)
+	   << " (max " << m_cct->_conf->osd_max_scrubs << ", remote " << m_scrubs_remote
+	   << ")" << dendl;
+
+  std::lock_guard l{resource_lock};
+  --m_scrubs_local;
+  ceph_assert(m_scrubs_local >= 0);
+}
+
+bool ScrubQueue::inc_scrubs_remote()
+{
+  std::lock_guard l{resource_lock};
+
+  if (m_scrubs_local + m_scrubs_remote < m_cct->_conf->osd_max_scrubs) {
+    dout(20) << __func__ << " " << m_scrubs_remote << " -> " << (m_scrubs_remote + 1)
+	     << " (max " << m_cct->_conf->osd_max_scrubs << ", local " << m_scrubs_local
+	     << ")" << dendl;
+    ++m_scrubs_remote;
+    return true;
+  }
+
+  dout(20) << __func__ << " " << m_scrubs_local << " local + " << m_scrubs_remote
+	   << " remote >= max " << m_cct->_conf->osd_max_scrubs << dendl;
+  return false;
+}
+
+void ScrubQueue::dec_scrubs_remote()
+{
+  dout(20) << __func__ << " " << m_scrubs_remote << " -> " << (m_scrubs_remote - 1)
```

Comment:
`m_scrubs_remote` is accessed outside the critical section. It's a logging code but this would introduce another degree of freedom for a developer during an investigation: _does the text in the file fit the internal state?_

Reply:
Not a clear-cut decision, as there is a benefit in shortening the time under lock, but I think you are right.  Will move the log back under the lock.